integration/teleterm_test.go

```diff
@@ -335,6 +342,77 @@ func testHeadlessWatcher(t *testing.T, pack *dbhelpers.DatabasePack, creds *help
 	)
 }
 
+func testRemoteClientCache(t *testing.T, pack *dbhelpers.DatabasePack, creds *helpers.UserCreds) {
+	t.Helper()
+	ctx := context.Background()
+
+	tc := mustLogin(t, pack.Root.User.GetName(), pack, creds)
+
+	storageFakeClock := clockwork.NewFakeClockAt(time.Now())
+
+	storage, err := clusters.NewStorage(clusters.Config{
+		Dir:                tc.KeysDir,
+		Clock:              storageFakeClock,
+		InsecureSkipVerify: tc.InsecureSkipVerify,
+	})
+	require.NoError(t, err)
+
+	cluster, _, err := storage.Add(ctx, tc.WebProxyAddr)
+	require.NoError(t, err)
+
+	daemonService, err := daemon.New(daemon.Config{
+		Storage: storage,
+		CreateTshdEventsClientCredsFunc: func() (grpc.DialOption, error) {
+			return grpc.WithTransportCredentials(insecure.NewCredentials()), nil
+		},
+		KubeconfigsDir: t.TempDir(),
+		AgentsDir:      t.TempDir(),
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		daemonService.Stop()
+	})
+
+	// Check if parallel calls trying to get a client will return the same one.
+	var wg sync.WaitGroup
+	concurrentCalls := 2
+	var parallelCallsForClient = make([]*client.ProxyClient, concurrentCalls)
+	for i := 0; i < concurrentCalls; i++ {
+		wg.Add(1)
+		go func(i int) {
```

Comment:
I saw people create a channel outside of the concurrent functions and then close it once both goroutines are spawned. I suppose it has a better shot at actually executing that code in parallel.

https://github.com/gravitational/teleport/blob/7ff5365b86775292c3ff0422d029a451dc61db70/lib/utils/fanoutbuffer/buffer_test.go#L226-L257

Reply:
+1, I'd also increase the amount of attempts a bit.

Reply:
Thanks, TIL again.